core/dbt/include/global_project/macros/materializations/models/incremental/strategies.sql

```diff
@@ -0,0 +1,83 @@
+{% macro get_incremental_append_sql(arg_dict) %}
+
+  {{ return(adapter.dispatch('get_incremental_append_sql', 'dbt')(arg_dict)) }}
+
+{% endmacro %}
+
+
+{% macro default__get_incremental_append_sql(arg_dict) %}
+
+  {% do return(get_insert_into_sql(arg_dict["target_relation"], arg_dict["temp_relation"], arg_dict["dest_columns"])) %}
+
+{% endmacro %}
+
+
+{# snowflake #}
+{% macro get_incremental_delete_insert_sql(arg_dict) %}
+
+  {{ return(adapter.dispatch('get_incremental_delete_insert_sql', 'dbt')(arg_dict)) }}
+
+{% endmacro %}
+
+{% macro default__get_incremental_delete_insert_sql(arg_dict) %}
+
+  {% do return(get_delete_insert_merge_sql(arg_dict["target_relation"], arg_dict["temp_relation"], arg_dict["unique_key"], arg_dict["dest_columns"])) %}
+
+{% endmacro %}
+
+
+{# snowflake, bigquery, spark #}
+{% macro get_incremental_merge_sql(arg_dict) %}
+
+  {{ return(adapter.dispatch('get_incremental_merge_sql', 'dbt')(arg_dict)) }}
+
+{% endmacro %}
+
+{% macro default__get_incremental_merge_sql(arg_dict) %}
+
+  {% do return(get_merge_sql(arg_dict["target_relation"], arg_dict["temp_relation"], arg_dict["unique_key"], arg_dict["dest_columns"])) %}
+
+{% endmacro %}
+
+
+{% macro get_incremental_insert_overwrite_sql(arg_dict) %}
+
+  {{ return(adapter.dispatch('get_incremental_insert_overwrite_sql', 'dbt')(arg_dict)) }}
+
+{% endmacro %}
+
+{% macro default__get_incremental_insert_overwrite_sql(arg_dict) %}
+
+  {% do return(get_insert_overwrite_merge_sql(arg_dict["target_relation"], arg_dict["temp_relation"], arg_dict["dest_columns"], arg_dict["predicates"])) %}
+
+{% endmacro %}
+
+
+{% macro get_incremental_default_sql(arg_dict) %}
+
+  {{ return(adapter.dispatch('get_incremental_default_sql', 'dbt')(arg_dict)) }}
+
+{% endmacro %}
+
+{% macro default__get_incremental_default_sql(arg_dict) %}
+
+  {% if arg_dict["unique_key"] %}
+    {% do return(get_incremental_delete_insert_sql(arg_dict)) %}
+  {% else %}
+    {% do return(get_incremental_append_sql(arg_dict)) %}
+  {% endif %}
+
+{% endmacro %}
```

Comment:
This is clever. I like it!

- We want similar behavior for `dbt-snowflake` + `dbt-bigquery`, though using `merge` instead of `delete+insert` if a `unique_key` is specified
- I think this should be the behavior for `dbt-postgres` (+ `dbt-redshift`), but I'm not sure it should be the true default (global project) behavior.
- I think `append` is more realistic as universal default. That's also the default we want for `dbt-spark`

Reply:
I 100% agree that should keep being the behavior for `dbt-postgres`. I'm wondering if what's above should be `default__get_incremental_default_sql` — or if that logic should live in `postgres__get_incremental_default_sql`, with the "true" default just:
```sql
{% macro default__get_incremental_default_sql(arg_dict) %}
    {% do return(get_incremental_append_sql(arg_dict)) %}
{% endmacro %}
```
Reason being: I'd like to see us move into a world where `default` != `postgres`, at least not exactly, even if PostgreSQL is still the reference implementation for most desirable behaviors.

Many many analytical databases don't support `delete+insert`, and I'm not sure we should take it as the high water-mark to which they should aspire. An adapter that _just_ supports `append` is pretty uninspiring, but it is the thing we test for today in the "basic" incremental test: [`dbt/tests/adapter/basic/test_incremental.py`](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/basic/test_incremental.py)